convert2rhel/systeminfo.py

```diff
@@ -152,49 +155,89 @@ def get_system_release_file_content():
 
-    def _get_system_version(self, system_release_content=None):
-        """Return a namedtuple with major and minor elements, both of an int type.
+        If system_release_content is not provided we use the content from the self.system_release_file_content.
 
-        Examples:
-        Oracle Linux Server release 7.8
-        CentOS Linux release 7.6.1810 (Core)
-        CentOS Linux release 8.1.1911 (Core)
+        :param system_release_content: The contents of the system_release file if needed.
+        :type system_release_content: str
+        :returns: A dictionary containing the system release information
+        :rtype: dict[str, str]
         """
+
         content = self.system_release_file_content if not system_release_content else system_release_content
-        match = re.search(r".+?(\d+)\.(\d+)\D?", content)
+
+        match = re.match(
+            # We assume that the /etc/system-release content follows the pattern:
+            # "<name> release <full_version> <Beta> (<dist_id>)"
+            # Here
+            # - <name>, parsed as a named group '(?P<name>.+?)',
+            #   is a non-empty string of arbitrary symbols.
+            # - "release ", parsed as (?:release\s)?,
+            #   is a literal string. It is optional, and it is dropped when parsed, so this group doesn't have a name.
+            # - <full_version>, parsed as a named group '(?P<full_version>[.\d]+)',
+            #   is a string of arbitrary length containing numbers and dots, for example 8.1.1911 or 7.9.
+            #   (For now we do not have examples with version strings containing letters or other symbols)
+            # - <Beta>, just the word Beta which may appear in some cases, parsed as '(?:\sBeta)?'.
+            # - <dist_id>, parsed as a named group (?P<dist_id>.+), is optional and when it is present, it must appear
+            #   in brackets. Thus, the named group is nested under an unnamed group which starts from a space and then
+            #   has brackets '(\s\( ...here goes the nested group... \))?'
+            #
+            # Example:
+            #
+            #   CentOS Stream release 8
+            #   <    name   > <      ><full_version>
+            #
+            #   CentOS Linux release 8.1.1911        (Core     )
+            #   <    name  > <      ><full_version><  <dist_id>>
+            #   CentOS Linux release 8.1.1911      Beta       (Core     )
+            #   <    name  > <      ><full_version><    ><  <dist_id>>
+            r"^(?P<name>.+?)\s(?:release\s)?(?P<full_version>[.\d]+)(?:\sBeta)?(\s\((?P<dist_id>.+)\))?$",
+            content,
+        )
+
         if not match:
-            self.logger.critical("Couldn't get system version from the content string: %s" % content)
-        version = Version(int(match.group(1)), int(match.group(2)))
+            self.logger.critical_no_exit("Couldn't parse the system release content string: %s" % content)
```

Comment:
Why do we changed this to no_exit? It seems to me very critical to inform that the system release file is not correct. 

Also, I think we will raise an exception while resolving system info, right? Is it fine if we just bubble up the exception?

Reply:
```
[2024-04-05T13:49:33+0000] TASK - [Prepare: Gather system information] *******************************
CRITICAL - Couldn't parse the system release content string: CentOS Linux release 

Writing breadcrumbs to '/etc/migration-results'.
[2024-04-05T13:49:33+0000] DEBUG - No RHSM facts folder found at '/etc/rhsm/facts'. Creating a new one...
Writing RHSM custom facts to '/etc/rhsm/facts/convert2rhel.facts'.
[2024-04-05T13:49:33+0000] DEBUG - Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/convert2rhel/main.py", line 108, in main_locked
    gather_system_info()
  File "/usr/lib/python3.6/site-packages/convert2rhel/main.py", line 263, in gather_system_info
    systeminfo.system_info.resolve_system_info()
  File "/usr/lib/python3.6/site-packages/convert2rhel/systeminfo.py", line 122, in resolve_system_info
    self.name = system_release_data["name"]
KeyError: 'name'

No changes were made to the system.
[2024-04-05T13:49:33+0000] DEBUG - /var/run/lock/convert2rhel.pid PID 34285 unlocked.
make: *** [Makefile:25: analyze] Error 1
```

Reply:
My intent is that parse_.. function is a simple utility function. It should not make decisions on should we stop the execution or not. If parses what it can find, and  then the higher-level structure may decide what to do with this information. In this case it still fails, and that's expected.

It maybe eventually we would be overriding these parameters from a config file or whatever.

I could have raise a custom exception here instead of passing None as a reply. But it seems like an overkill.

Reply:
Alright, got the idea behind... And I agree that `parse_*` functions should all have that same intention, but at the same time, I feel that we are making some type of "regression". Right now, when this fails, it will show to the user a red message with the exact error, making it less scary if something goes wrong. 

With this change, we are going away from a message that seems informative about the error, and just raising an base exception that, sometimes, may not be useful to the user. 

What if we handle this type of exception in a higher level function, just so we can maintain the same UX for the user, without them noticing that we changed how we parse the content string?